File: src/main/java/org/demo/domain/BankAccount.java
package org.demo.domain;

import lombok.NoArgsConstructor;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventhandling.EventHandler;
import org.axonframework.eventhandling.GenericEventMessage;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.AggregateIdentifier;
import org.axonframework.modelling.command.CreationPolicy;

import java.math.BigDecimal;

import static org.axonframework.modelling.command.AggregateLifecycle.apply;

@NoArgsConstructor
public class BankAccount {
	@AggregateIdentifier
	private String accountId;
	private String accountNumber;
	private BigDecimal balance;

	@CommandHandler
	@CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
	public void openAccount(OpenAccount command) throws AccountAlreadyOpenedException {
		if (accountId != null) {
			throw new AccountAlreadyOpenedException(accountId);
		}
		apply(new AccountOpened(command.accountId(), command.accountNumber()));
	}

	@EventHandler
	public void accountOpened(AccountOpened event) {
		this.accountId = event.accountId();
		this.accountNumber = event.accountNumber();
		this.balance = BigDecimal.valueOf(0);
	}

	@CommandHandler
	public void depositMoney(DepositMoney command) {
		apply(new MoneyDeposited(
				command.accountId(),
				command.amount(),
				GenericEventMessage.clock.instant()));
	}

	@EventHandler
	public void moneyDeposited(MoneyDeposited event) {
		this.balance = this.balance.add(event.amount());
	}

	@CommandHandler
	public void withdrawMoney(WithdrawMoney command) throws OverdraftDetectedException {
		if (balance.compareTo(command.amount()) >= 0) {
			apply(new MoneyWithdrawn(
					command.accountId(),
					command.amount(),
					GenericEventMessage.clock.instant()));
		} else {
			throw new OverdraftDetectedException(accountNumber, balance, command.amount());
		}
	}

	@EventHandler
	public void moneyWithdrawn(MoneyWithdrawn event) {
		this.balance = this.balance.subtract(event.amount());
	}
}

File: src/main/kotlin/org/demo/domain/BankAccount.kt
package org.demo.domain

import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventhandling.EventHandler
import org.axonframework.eventhandling.GenericEventMessage
import org.axonframework.modelling.command.AggregateCreationPolicy
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle.apply
import org.axonframework.modelling.command.CreationPolicy
import org.axonframework.spring.stereotype.Aggregate
import java.math.BigDecimal

@Aggregate
class BankAccount() {
	@AggregateIdentifier
	private lateinit var accountId: String
	private lateinit var accountNumber: String
	private lateinit var balance: BigDecimal
	private var accountStatus: AccountStatus = AccountStatus.NEW

	enum class AccountStatus {
		NEW,
		OPENED,
		CLOSED
	}

	@CommandHandler
	@CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
	@Throws(AccountAlreadyOpenedException::class)
	fun openAccount(command: OpenAccount) {
		if (accountStatus != AccountStatus.NEW) {
			throw AccountAlreadyOpenedException(accountId)
		}
		apply(AccountOpened(command.accountId, command.accountNumber))
	}

	@EventHandler
	fun accountOpened(event: AccountOpened) {
		accountStatus = AccountStatus.OPENED
		accountId = event.accountId
		accountNumber = event.accountNumber
		balance = BigDecimal.valueOf(0)
	}

	@CommandHandler
	fun closeAccount(command: CloseAccount) {
		apply(AccountClosed(command.accountId))
	}

	@EventHandler
	fun accountClosed(event: AccountClosed) {
		accountStatus = AccountStatus.CLOSED;
	}

	@CommandHandler
	fun depositMoney(command: DepositMoney) {
		apply(MoneyDeposited(
				command.accountId,
				command.amount,
				GenericEventMessage.clock.instant()))
	}

	@EventHandler
	fun moneyDeposited(event: MoneyDeposited) {
		balance = balance.add(event.amount)
	}

	@CommandHandler
	@Throws(OverdraftDetectedException::class)
	fun withdrawMoney(command: WithdrawMoney) {
		if (balance.compareTo(command.amount) >= 0) {
			apply(MoneyWithdrawn(
					command.accountId,
					command.amount,
					GenericEventMessage.clock.instant()))
		} else {
			throw OverdraftDetectedException(accountNumber, balance, command.amount)
		}
	}

	@EventHandler
	fun moneyWithdrawn(event: MoneyWithdrawn) {
		balance = balance.subtract(event.amount)
	}
}
